src/components/Header.tsx

```diff
@@ -0,0 +1,146 @@
+/* eslint-disable @typescript-eslint/indent */
+/* eslint-disable @typescript-eslint/no-unused-expressions */
+import { useEffect, useRef, useState } from 'react';
+import { Todo } from '../types/Todo';
+import { addTodo, updateTodo } from '../api/todos';
+
+interface HeaderProps {
+  todos: Todo[];
+  setEditingTodosId: (ids: number[]) => void;
+  setError: (error: string) => void;
+  editingTodosId: number[];
+  setTodos: (todos: Todo[]) => void;
+  setFilteredTodos: (todos: Todo[]) => void;
+  setIsActiveFilter: (isActive: boolean) => void;
+  filteredTodos: Todo[];
+}
+
+export function Header({
+  todos,
+  setEditingTodosId,
+  setTodos,
+  setFilteredTodos,
+  setError,
+  editingTodosId,
+}: HeaderProps) {
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+  const [isAddingNewTodo, setIsAddingNewTodo] = useState(false);
+
+  const inputNewTodo = useRef<HTMLInputElement>(null);
+
+  function handleNewTodoTitle(e: React.ChangeEvent<HTMLInputElement>) {
+    setNewTodoTitle(e.target.value);
+  }
+
+  async function handleAddNewTodo(e: React.FormEvent<HTMLFormElement>) {
+    e.preventDefault();
+
+    if (!newTodoTitle.trim()) {
+      setError('Title should not be empty');
+      inputNewTodo.current?.focus();
+
+      return;
+    }
+
+    const newTodoIdAvailable = todos.length
+      ? todos[todos.length - 1].id + 1
+      : 1;
+
+    try {
+      setIsAddingNewTodo(true);
+      setEditingTodosId([...editingTodosId, newTodoIdAvailable]);
+
+      const newTodo = {
+        title: newTodoTitle.trim(),
+        completed: false,
+        id: newTodoIdAvailable,
+      };
+
+      setFilteredTodos([...todos, newTodo as Todo]);
+
+      const responseFromAddedTodo = await addTodo(newTodoTitle.trim());
+
+      setFilteredTodos([...todos].filter(t => t.id !== newTodoIdAvailable));
+      setTodos([...todos, responseFromAddedTodo]);
+      setFilteredTodos([...todos, responseFromAddedTodo]);
+      setNewTodoTitle('');
+    } catch (er) {
+      setEditingTodosId(
+        [...editingTodosId].filter(id => id !== newTodoIdAvailable),
+      );
+      setError('Unable to add a todo');
+      setFilteredTodos([...todos].filter(t => t.id !== newTodoIdAvailable));
+    } finally {
+      setEditingTodosId(editingTodosId.filter(id => id !== newTodoIdAvailable));
+
+      setIsAddingNewTodo(false);
+    }
+  }
+
+  function handleToggleAllTodos() {
+    const successfulUpdatedTodo: number[] = [];
+    const hasAllSameStatus = todos.every(t => t.completed);
```

Comment:
This is not actually a mistake, just an possible improvement.

add a `Complete` to the name of the variable. You want it to have the value `true` if all the `todos` have status `completed`. When you name the varibale `hasAllSameStatus` it gives the impression that if the status of all the `todos` is `not completed`, it would be true as well.